File: programmers/kakao2020_stringCompression.js
/*
	rank:	1
	title:	[2020카카오공채] 문자열 압축
	link: 	https://programmers.co.kr/learn/courses/30/lessons/60057
*/

function solution(inputStr) {
    let list = [inputStr];

    for (let i = 1; i <= Math.floor(inputStr.length / 2); i++) {
        let count = 1;
        let output = "";
        let prevStr = "";

        for (let j = 0; j < inputStr.length / i; j++) {
            let subString = inputStr.substr(j * i, i);

            if (prevStr == subString) {
                count++;
            } else {
                if (count >= 2) {
                    output += count;
                }
                output += prevStr;
                count = 1;
            }

            prevStr = subString;
        }

        if (count >= 2) {
            output += count;
        }
        output += prevStr;

        list.push(output);
    }

    return Math.min(...list.map(str => str.length));
}

File: programmers/etc/python/kakao2020_stringCompression.py
# rank:	    1
# title:	[2020카카오공채] 문자열 압축
# link: 	https://programmers.co.kr/learn/courses/30/lessons/60057

import math
from functools import cmp_to_key

def custom_compare(a, b):
    return len(a) - len(b)

def solution(s):
    list = [s]

    for i in range(1, math.floor(len(s) / 2) + 1):
        prevLetter = ""
        count = 1
        item = ""

        for j in range(0, len(s), i):
            if prevLetter == s[j:j+i]:
                count += 1
            else:
                if count >= 2:
                    item += str(count)
                item += prevLetter
                count = 1
            prevLetter = s[j:j+i]

        if count >= 2:
            item += str(count)
        item += prevLetter
        list.append(item)

    list.sort(key=cmp_to_key(custom_compare))
    return len(list[0])